### PR TITLE
Fix #16078 - add type to CMakePresets cacheVariables

### LINKED ISSUE #16078: [feature] CMakeToolchain cacheVariables should have a type
### What is your suggestion?

Type mismatch in CMake configure arguments is a common issue that is really hard to track down, and as such I'm really pedantic about specifying the type of the variables when configuring CMake, especially for path/filepath ones.

```diff
--DMY_FLAG=ON
+-DMY_FLAG:BOOL=ON

-DMY_PATH=/path/to/file
+DMY_PATH:FILEPATH=/path/to/file

-DMY_DIR_PATH=/path/to/dir
+DMY_DIR_PATH:PATH=/path/to/dir
```

I noticed that the CMake presets do not hold a type for `cacheVariables`, while the CMake documentation explicitly says it can be a value, or a an object with both `type` and `value`, cf https://cmake.org/cmake/help/v3.29/manual/cmake-presets.7.html#configure-preset

Perhaps should consider explictly setting it, around https://github.com/conan-io/conan/blob/f1ef6e461accc293b14a7bf883865858ac29c7ad/conan/tools/cmake/toolchain/toolchain.py#L240-L271

Related:

* #14590
* #14568

---- 

Quoting Professional CMake - Practical Guide, 15th edition by Craig Scott (I hope that's ok, since I can already find the same online at https://discourse.cmake.org/t/cmake-d-var-type-value/1135)

> There is a special case for handling values initially declared without a type on the cmake command
line. If the project’s CMakeLists.txt file then tries to set the same cache variable and specifies a type
of FILEPATH or PATH , then if the value of that cache variable is a relative path, CMake will treat it as
being relative to the directory from which cmake was invoked and automatically convert it to an
absolute path. This is not particularly robust, since cmake could be invoked from any directory, not
just the build directory. Therefore, developers are advised to always include a type if specifying a
variable on the cmake command line for a variable that represents some kind of path. 

> It is a good habit to always specify the type of the variable on the command line in general anyway so that it is
likely to be shown in GUI applications in the most appropriate form. It will also prevent one of the
surprising behavior scenarios mentioned in the previous section

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

### PR DESCRIPTION
**Pull request by jmarrec**
Changelog: Feature: add type to CMakePresets cacheVariables

Close #16078

* Fix #16078 
* I am assuming I'm missing tests and co, but this is more a proof of concept than anything else, to show that it wouldn't be too much effort to implement
* Add types for BOOL, STRING, FILEPATH/PATH (based on assigning a `pathlib.Path` to the `tc.cache_variables`)

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop2/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one.

### CODE REVIEW

File: conan/tools/cmake/toolchain/toolchain.py [hunk 1]
```diff
@@ -240,14 +241,21 @@ def generate(self):
...
+                cache_variables[name] = {"value": value, "type": "STRING"}
             elif isinstance(value, _PackageOption):
                 if str(value).lower() in ["true", "false", "none"]:
-                    cache_variables[name] = "ON" if bool(value) else "OFF"
+                    cache_variables[name] = {"value": "ON" if bool(value) else "OFF", "type": "BOOL"}
                 elif str(value).isdigit():
                     cache_variables[name] = int(value)
                 else:
-                    cache_variables[name] = str(value)
+                    cache_variables[name] = {"value": str(value), "type": "STRING"}
```
Comment (jmarrec) [inline conan/tools/cmake/toolchain/toolchain.py:258]: Replace cache_variables with an object with both `value` and `type` where appropriate

Test example: 

```
tc = CMakeToolchain(self)
tc.cache_variables["MY_BOOL"] = False
tc.cache_variables["MY_FILE_PATH"] = Path("/path/to/file")   # assert Path("/path/to/file").is_file()
tc.cache_variables["MY_FILE_PATH"] = Path("/path/to/dir")   # assert Path("/path/to/file").is_dir()
tc.generate()
```

File: conan/tools/cmake/presets.py [hunk 2]
```diff
@@ -44,7 +44,7 @@ def generate(conanfile, toolchain_file, generator, cache_variables, preset_prefi
 
         if "BUILD_TESTING" not in cache_variables:
             if conanfile.conf.get("tools.build:skip_test", check_type=bool):
-                cache_variables["BUILD_TESTING"] = "OFF"
+                cache_variables["BUILD_TESTING"] = {"value": "OFF", "type": bool}
```
Comment (jmarrec) [inline conan/tools/cmake/presets.py]: Also add it to a few vars in presets.py itself

File: conan/tools/cmake/presets.py [hunk 3]
```diff
@@ -158,12 +158,22 @@ def _configure_preset(conanfile, generator, cache_variables, toolchain_file, mul
...
+                value = value_or_dict
+
+            return f"-D{var}{type_str}={_format_val(value)}"
 
         # https://github.com/conan-io/conan/pull/12034#issuecomment-1253776285
         cache_variables_info = " ".join(
-            [f"-D{var}={_format_val(value)}" for var, value in cache_variables.items()])
-        add_toolchain_cache = f"-DCMAKE_TOOLCHAIN_FILE={toolchain_file} " \
+            [_format_var(var, value) for var, value in cache_variables.items()])
+        add_toolchain_cache = f"-DCMAKE_TOOLCHAIN_FILE:FILEPATH={toolchain_file} " \
```
Comment (jmarrec) [inline conan/tools/cmake/presets.py:182]: Properly format the message that is printed as well:

```
conanfile.py: If your CMake version is not compatible
with CMakePresets (<3.23) call cmake like: 
'cmake <path> -G Ninja 
 -DCMAKE_TOOLCHAIN_FILE:FILEPATH=/media/path/to/conan_toolchain.cmake
 -DMY_BOOL:BOOL=ON 
 -DMY_FILE_PATH:FILEPATH=/path/to/file 
 -DMY_DIR_PATH:PATH=/path/to/dir
```

**Comment by memsharded** [comment]
I have done a commit to fix one broken "integration" test, but then a lot of "functional" tests are failing:

[logs.txt](https://github.com/conan-io/conan/files/14998113/logs.txt) (Linux logs, but other OSs are failing as well)

And it is not failing in obvious ways. So this is a big concern, that this change might be breaking users in unexpected ways.
To move this forward it will be necessary a very compelling example of something that clearly fails (and can be marked as a bug) when the type of the cache variable is not specified, because so far it has been working quite smoothly.

**Comment by memsharded** [comment]
Hi @jmarrec 

Have you been able to check my comment above? 

> To move this forward it will be necessary a very compelling example of something that clearly fails (and can be marked as a bug) when the type of the cache variable is not specified, because so far it has been working quite smoothly.

The changes you are proposing are not risk-free, even if the broken tests might be fixed, the risk of breaking users is still there. So it would be necessary to provide some broken/bug cases in which the proposed changes solve the issue.

Can you please try to provide such a "red" test to re-consider this PR? Otherwise I think we will have to discard this PR, even if the typing is a nice-to-have, if it is not solving some actual user issues, it might not be worth the risk of breaking users.

**Comment by jmarrec** [comment]
I can't think of a way to produce a bug, so if you don't want to pursue it that's fine.

**Comment by memsharded** [comment]
Yes, I think that unless it is solving some specific issue that we can reproduce, it would be better not to add this extra complexity, as it doesn't seem completely risk-free. Thanks for understanding.

**Pull request closed without merging**